Run TF-PSA-Crypto components without cloning Mbed TLS

bensze01 opened the pull request:
Fixes Mbed-TLS/TF-PSA-Crypto#116

- Internal CI:
  - mbed-tls-pr-test-parametrized
    - [Mbed TLS + TF-PSA-Crypto][17]
    - [Mbed TLS][18]
    - [Framework CI + TF-PSA-Crypto][19]
  - mbedtls-release-ci-testing
    - [Mbed TLS + TF-PSA-Crypto][20]
    - [Mbed TLS][21]
    - [Framework CI + TF-PSA-Crypto][22]
- OpenCI:
  - mbed-tls-restricted-pr-test-parametrized
    - [Mbed TLS + TF-PSA-Crypto][23]
    - [Mbed TLS][24]
    - [Framework CI + TF-PSA-Crypto][25]
  - mbedtls-release-ci-testing
    - [Mbed TLS + TF-PSA-Crypto][26]
    - [Mbed TLS][27]
    - [Framework CI + TF-PSA-Crypto][28]
      
 Older runs:
- 1cfc68d72a9874e3c82762404710fd7cbc67e264
  - Internal CI:
    - mbed-tls-pr-test-parametrized
      - [Mbed TLS + TF-PSA-Crypto][13]
      - [Mbed TLS][14]
    - mbedtls-release-ci-testing
      - [Mbed TLS + TF-PSA-Crypto][15]
      - [Mbed TLS][16]
  - OpenCI:
    - mbed-tls-restricted-pr-test-parametrized
      - [Mbed TLS + TF-PSA-Crypto][9]
      - [Mbed TLS][10]
    - mbedtls-release-ci-testing
      - [Mbed TLS + TF-PSA-Crypto][11]
      - [Mbed TLS][12]
- 9b821846659e29906101e2e5ad891cde2ce60fc6
  - Internal CI:
    - mbed-tls-pr-test-parametrized
      - [Mbed TLS + TF-PSA-Crypto][1]
      - [Mbed TLS][2]
    - mbedtls-release-ci-testing
      - [Mbed TLS + TF-PSA-Crypto][3]
      - [Mbed TLS][4]
  - OpenCI:
    - mbed-tls-restricted-pr-test-parametrized
      - [Mbed TLS + TF-PSA-Crypto][5]
      - [Mbed TLS][6]
    - mbedtls-release-ci-testing
      - [Mbed TLS + TF-PSA-Crypto][7]
      - [Mbed TLS][8]

[1]: https://jenkins-mbedtls.oss.arm.com/job/mbed-tls-pr-test-parametrized/413/
[2]: https://jenkins-mbedtls.oss.arm.com/job/mbed-tls-pr-test-parametrized/414/
[3]: https://jenkins-mbedtls.oss.arm.com/job/mbedtls-release-ci-testing/753/
[4]: https://jenkins-mbedtls.oss.arm.com/job/mbedtls-release-ci-testing/754/
[5]: https://mbedtls.trustedfirmware.org/job/mbed-tls-restricted-pr-test-parametrized/82/
[6]: https://mbedtls.trustedfirmware.org/job/mbed-tls-restricted-pr-test-parametrized/83/
[7]: https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/317/
[8]: https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/318/

[9]: https://mbedtls.trustedfirmware.org/job/mbed-tls-restricted-pr-test-parametrized/97/
[10]: https://mbedtls.trustedfirmware.org/job/mbed-tls-restricted-pr-test-parametrized/98/
[11]: https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/326/
[12]: https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/327/
[13]: https://jenkins-mbedtls.oss.arm.com/job/mbed-tls-pr-test-parametrized/450/
[14]: https://jenkins-mbedtls.oss.arm.com/job/mbed-tls-pr-test-parametrized/451/
[15]: https://jenkins-mbedtls.oss.arm.com/job/mbedtls-release-ci-testing/764/
[16]: https://jenkins-mbedtls.oss.arm.com/job/mbedtls-release-ci-testing/765/

[17]: https://jenkins-mbedtls.oss.arm.com/job/mbed-tls-pr-test-parametrized/453/
[18]: https://jenkins-mbedtls.oss.arm.com/job/mbed-tls-pr-test-parametrized/454/
[19]: https://jenkins-mbedtls.oss.arm.com/job/mbed-tls-pr-test-parametrized/456/
[20]: https://jenkins-mbedtls.oss.arm.com/job/mbedtls-release-ci-testing/783/
[21]: https://jenkins-mbedtls.oss.arm.com/job/mbedtls-release-ci-testing/781/
[22]: https://jenkins-mbedtls.oss.arm.com/job/mbedtls-release-ci-testing/784/
[23]: https://mbedtls.trustedfirmware.org/job/mbed-tls-restricted-pr-test-parametrized/103/
[24]: https://mbedtls.trustedfirmware.org/job/mbed-tls-restricted-pr-test-parametrized/104/
[25]: https://mbedtls.trustedfirmware.org/job/mbed-tls-restricted-pr-test-parametrized/106/
[26]: https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/346/
[27]: https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/347/
[28]: https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/348/

mpg commented on vars/common.groovy:
```diff
@@ -256,26 +262,33 @@ List<BranchInfo> get_branch_information(Collection<String> branches) {
+                                    ),
+                                    returnStdout: true
+                                ).trim().split('\n')
+                                echo "All all.sh components: ${all.join(" ")}"
```
Minor should probably adjust the message as well: All all.sh -> <repo> all.sh?

mpg commented on vars/gen_jobs.groovy:
```diff
@@ -245,7 +245,7 @@ echo >&2 'Note: "clang" will run /usr/bin/clang -Wno-error=c11-extensions'
     }
 
-    if (info.has_min_requirements) {
+    if (repo == 'tls' && info.has_min_requirements) {
```
I'm not sure I understand why we need the `repo == 'tls'` part: currently `tf-psa-crypto` doesn't have `scripts/min_requirements.py` but I'd guess `info.has_min_requirements` should reflect that?

mpg replied:
Also, it seems to me we'll need to get `scripts/min_requirements.py` (and supporting files) in tf-psa-crypto (or the framework) before moving python scripts that have requirements beyond what happens to be already installed by apt. @ronald-cron-arm not sure if that dependency was already on your radar?

ronald-cron-arm replied:
No that was not on my radar. I can see that @valeriosetti has started working on https://github.com/Mbed-TLS/mbedtls-framework/issues/86. Does this issue fully cover the dependency?

valeriosetti replied:
> I can see that @valeriosetti has started working on https://github.com/Mbed-TLS/mbedtls-framework/issues/86.

Yes, PRs https://github.com/Mbed-TLS/mbedtls-framework/pull/105 and related ones (https://github.com/Mbed-TLS/mbedtls/pull/9863 https://github.com/Mbed-TLS/mbedtls/pull/9864) are ready for review. There's only the usual `framework` confict on `development` and `mbedtls-3.6`, but that can easily be fixed once https://github.com/Mbed-TLS/mbedtls-framework/pull/105 is merged.

bensze01 replied:
@mpg Currently the code only checks if `min_requirements.py` exists in the MbedTLS branch we are testing, and this is what's represented by `has_min_requirements`. Since we're planning to add such a script, I'll add a separate variable to track this for the framework.

valeriosetti replied:
For the records, I created https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/148 to add `min_requirements.py` to TF-PSA-Crypto

mpg replied:
> Since we're planning to add such a script, I'll add a separate variable to track this for the framework.

Not sure if this was a typo, but just to be sure: I was talking about tf-psa-crypto, not the framework. Currently the plan is for `scripts/min_requirements.py` to exist only in the two "top-level" repos (mbedtls, tf-psa-crypto).

bensze01 replied:
Ah, yes sorry, that was a typo - I meant tf-psa-crypto.

mpg commented on vars/gen_jobs.groovy:
```diff
@@ -258,6 +258,16 @@ scripts/min_requirements.py --user ${info.python_requirements_override_file}
+                    default:
+                        error("Invalid repo: $repo")
+                }
                 checkout_repo.checkout_tls_repo(info)
```
Don't you mean to remove this line?

mpg commented on vars/gen_jobs.groovy:
```diff
@@ -668,6 +678,12 @@ def gen_release_jobs(BranchInfo info, String label_prefix='', boolean run_exampl
 
+    if (env.RUN_TF_PSA_CRYPTO_ALL_SH == "true") {
+        info.tf_psa_crypto_all_sh_components.each({ component, platform ->
+            jobs = jobs + gen_all_sh_jobs(info, platform, component, label_prefix)
```
Shouldn't we pass `'tf-psa-crypto'` as the last argument? Otherwise it will default to `'tls'`.

mpg reviewed (commented):
Just making myself familiar with the code.

Note: we'll want a test run showing that if a tf-psa-crypto component tries to reference an Mbed TLS file, it fails. (We'll probably want to create a faulty PR for that purpose.)

bensze01 commented:
@mpg I think I've addressed all the feedback. 
Adding support for `min_requirements.py` required a major refactor that brought the design back closer to the way `get_branch_information()` was set up, so I've squashed the back-and-forth changes to hopefully make reviewing this PR easier.

bensze01 commented:
All the test runs completed successfully.

ronald-cron-arm commented on vars/common.groovy:
```diff
@@ -224,73 +224,52 @@ docker run -u \$(id -u):\$(id -g) -e MAKEFLAGS -e VERBOSE_LOGS $env_args --rm --
  * test environment.
  * In particular, get components of all.sh for Linux platforms. */
-List<BranchInfo> get_branch_information(Collection<String> branches) {
+List<BranchInfo> get_branch_information(Collection<String> tls_branches) {
```
As discussed offline it seems it would be better to have another argument here with the tf-psa-crypto branches that would bubble up into `run_pr_job()` signature as well rather than adding here a `tf-psa-crypto` branch based on `env.RUN_TF_PSA_CRYPTO_ALL_SH`. This should help as well when we add the TF-PSA-Crypto development branch as a branch to be tested as part of the framework CI.

ronald-cron-arm commented:
@bensze01 have you run new tests with the last version?

yanesca reviewed (commented):
I left a couple of remarks inline.

Also, commit "Test tf-psa-crypto components without Mbed TLS" a lot of things are happening:
1. Adds a function for unified handling of repo checkout
2. Adds another level of loop for dealing with multiple repos when collecting branch information
3. Restricts non-all.sh jobs to the tls repo (with the exception of analyse outcomes which follows in the next commit)
4. Extends prefix handling to handle multiple repos
I am not asking for any changes here, just for future reference, in similar
cases breaking something like this down into more commits helps review and future development.

Overall I didn't find any blockers, the only reason I am not approving yet because I have a question about the test runs:
The reason for failure seems unrelated to this PR, still it makes me wonder if we understand what is going on there and why we don't want to get a green CI?

yanesca reviewed (approved):
> the only reason I am not approving yet because I have a question about the test runs

Oh, never mind, it is the ongoing CI incident.

bensze01 commented on vars/mbedtls.groovy:
```diff
@@ -64,7 +64,7 @@ void run_tls_tests(Collection<BranchInfo> infos) {
 void run_pr_job(String target_repo, boolean is_production, String tls_branches, String tf_psa_crypto_branches) {
     def (tls_split,    tf_psa_crypto_split) =
-        [tls_branches, tf_psa_crypto_branches]*.split(',')*.findAll()
+        [tls_branches, tf_psa_crypto_branches].collect({branches -> branches.split(',').findAll()})
```
I'm using `findAll()` instead of `as List` to filter out any empty strings returned by `.split(',')`.
Eg.:
```
''.split(',') as List == [ '' ]
''.split(',').findAll() == []
```

bensze01 commented on vars/mbedtls.groovy:
```diff
@@ -62,11 +62,13 @@ void run_tls_tests(Collection<BranchInfo> infos) {
 }
 
-void run_pr_job(String target_repo, boolean is_production, List<String> branches) {
+void run_pr_job(String target_repo, boolean is_production, Collection<String> tls_branches, Collection<String> tf_psa_crypto_branches) {
```
`Object[].findAll()` returns a `Collection` in its method signature (really an `ArrayList`, but that's an implementation detail)

bensze01 commented:
@mpg @yanesca I've rebased the PR on top of `main`, and all the CI runs passed.

yanesca reviewed (dismissed):
Did the rebase and arrived at the same result.

ronald-cron-arm reviewed (changes requested):
This looks almost good to me. I have a few comments though.

bensze01 commented on vars/mbedtls.groovy:
```diff
@@ -152,31 +161,40 @@ void run_pr_job(String target_repo, boolean is_production, List<String> branches
 
 void run_framework_pr_job() {
-    run_pr_job('framework', true, ['development', 'mbedtls-3.6'])
+    run_pr_job('framework', true, ['development', 'mbedtls-3.6'], ['development'])
```
It is possible to test this through `mbed-tls-restricted-pr-test-parametrized`, [by specifying multiple branches separated by commas](https://github.com/Mbed-TLS/mbedtls-test/blob/1cfc68d72a9874e3c82762404710fd7cbc67e264/vars/mbedtls.groovy#L65-L69).

ronald-cron-arm replied:
Ah great, could you test it then?

bensze01 commented:
All the tests (including the framework CI ones) have passed, please re-review, @yanesca @ronald-cron-arm.

ronald-cron-arm reviewed (approved):
This looks good to me, thanks.

yanesca reviewed (approved):
LGTM
